### PR TITLE
[Distributed] add distributed in-degree and out-degree.

### PR DESCRIPTION
**Pull request by zheng-da**
## Description
Add `in_degrees` and `out_degrees` in `DistGraph`.
This PR also fixes the issue that the out-degree computed from the partitioned graph is incorrect. To fix this issue, we need to add out-going edges between partitions on the source nodes.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/distributed/dist_graph.py [hunk 1]
```diff
@@ -745,6 +747,104 @@ def num_edges(self, etype=None):
                 return sum([self._gpb._num_edges(etype) for etype in self.etypes])
         return self._gpb._num_edges(etype)
 
+    def out_degrees(self, u=ALL):
+        """Return the out-degree(s) of the given nodes.
+
+        It computes the out-degree(s) w.r.t. to the edges of the given edge type.
```
Comment (classicsong) [inline python/dgl/distributed/dist_graph.py:753]: So where is the edge type?

File: src/graph/transform/partition_hetero.cc [hunk 2]
```diff
@@ -98,6 +98,32 @@ HaloHeteroSubgraph GetSubgraphWithHalo(std::shared_ptr<HeteroGraph> hg,
     }
   }
 
+  if (num_hops > 0) {
+    EdgeArray out_edges = hg->OutEdges(0, nodes);
```
Comment (classicsong) [inline src/graph/transform/partition_hetero.cc:102]: This is going to add halo nodes for out edges?

**Review by classicsong — APPROVED** [review]
A general question: when we add out-edges in each partitation for out_degree(), what is the memory and storage cost of doing so?

Overall, the PR LGTM.

File: src/graph/transform/partition_hetero.cc (hunk 2 above)
Comment (zheng-da) [reply src/graph/transform/partition_hetero.cc]: yes

**Comment by zheng-da** [comment]
this will save the out-edges that are cut by the partitioning algorithm with a partition. if METIS has many edge cuts, it shouldn't cause too much edge duplication.

**Review by classicsong — APPROVED** [review]
LGTM

**Pull request merged**